core/testing.go

```diff
@@ -52,26 +52,16 @@ func CreateKVStore(retainBlocks int64) *kvstore.Application {
 	return app
 }
 
-// StartTestClient returns a started remote Core node process, as well its
-// mock Core Client.
-func StartTestClient(ctx context.Context, t *testing.T) (tmservice.Service, Client) {
-	nd, _, cfg := StartTestKVApp(ctx, t)
-	endpoint, err := GetEndpoint(cfg)
-	require.NoError(t, err)
-	ip, port, err := net.SplitHostPort(endpoint)
-	require.NoError(t, err)
-
-	client, err := NewRemote(ip, port)
-	require.NoError(t, err)
-
-	err = client.Start()
-	require.NoError(t, err)
-	t.Cleanup(func() {
-		err := client.Stop()
-		require.NoError(t, err)
-	})
-
-	return nd, client
+func getFreePort() (port int, err error) {
+	var a *net.TCPAddr
+	if a, err = net.ResolveTCPAddr("tcp", "localhost:0"); err == nil {
+		var l *net.TCPListener
+		if l, err = net.ListenTCP("tcp", a); err == nil {
+			defer l.Close()
+			return l.Addr().(*net.TCPAddr).Port, nil
+		}
+	}
+	return
```

Comment:
Doesn't the zero port allocate the free one on the kernel? Why do we need to do it manually?

Reply:
Later in the code there is a [GetEndpoint](https://github.com/celestiaorg/celestia-node/blob/main/core/testing.go#L116-L128) which is used to build the test client.

```
func GetEndpoint(cfg *config.Config) (string, error) {
	url, err := url.Parse(cfg.RPC.ListenAddress)
	if err != nil {
		return "", err
	}
	host, _, err := net.SplitHostPort(url.Host)
	if err != nil {
		return "", err
	}
	return fmt.Sprintf("%s:%s", host, url.Port()), nil
}
```

So I need to know which port the kernel gave the test in order to build the  client.  
I do this [here](https://github.com/celestiaorg/celestia-node/pull/1400/files#diff-9750405595d72572f9a78b11e628a3cf4331798bb84dcd88746e61a8ba215657L91)
```
	// get a random port for running test in parallel
	freePort, err := getFreePort()
	require.NoError(t, err)
	tmNode.Config().RPC.ListenAddress = fmt.Sprintf("tcp://127.0.0.1:%d", freePort)
	freePort, err = getFreePort()
	require.NoError(t, err)
	tmNode.Config().P2P.ListenAddress = fmt.Sprintf("tcp://0.0.0.0:%d", freePort)

```

Reply:
Ok, I see. So we need RPC to get a port and create a client. Why do we need to get a free port for a P2P listen-to address, though? @dougEfresh

Reply:
Yeah, so let's just set it to 0

Reply:
and for RPC we can use a freePort

Reply:
But actually, this is a very minor thing to worry about, and using getFreePort is not that bad.
I just don't really like doing what kernel already provides, but its fine in this case

Reply:
Ah, I see what you mean. Minor thing yes, but I changed p2p port to use port 0 and tests pass.